### PR TITLE
Nz/3354 cost allocation validation

### LINKED ISSUE #3354: Add inline validation for Cost Allocation
Add inline validation for Cost Allocation

*These validations will be triggered by the error check that is currently being tested. #3488

|Field|Content|Considerations|
|---|---|---|
|Description of cost allocation methodology|Provide a description of the cost allocation methodology.|Should not be blank|
|Other funding amount|Provide an other funding amount greater than or equal to $0.|Should not be blank|
|Other funding description||Should be handled in #3439|

**Mockup:**. (How error text would fall, underneath the field.)

### This task is done when...

- [x] Description of cost allocation methodology has inline validation
- [ ] Other funding amount has inline validation

### PR DESCRIPTION
**Pull request by Sun-Mountain**
Resolves #3354 and #3547 

### Description
Validation for the CostAllocate and OtherFunding forms. Originally this was split into two tickets, however it wasn't divided by component but rather by function. Because #3547 was about the OtherFunding component, I decided to combine the two into one PR.

### Significant changes or possible side effects
Due to the nature of the dependent fields in the OtherFunding component, there is currently no react-hook controller around otherSources field, which may cause problems in the future with the admin check.

### Steps to manually verify this change

1. Go to an apd > activities > cost allocation
2. Methodology should be required
3. Other source should only be required if the corresponding other funds is greater than $0

### This pull request is ready to review when

- [x] Automated tests are updated (and all tests are passing)
- [ ] Pull request has been labeled, if applicable with feature, content, bug, tests, refactor
- [ ] Associated OpenAPI documentation has been updated
- [ ] The experience passes a basic manual accessibility audit (keyboard nav, screenreader, text scaling) OR an exemption is documented

### This pull request can be merged when

- [ ] Code has been reviewed by someone other than the original author
- [ ] QA has verified the accessibility and functionality related to the change
- [ ] Design has approved the experience
- [ ] Product has approved the experience

### CODE REVIEW

File: web/src/pages/apd/activities/cost-allocation/CostAllocation/OtherFunding.js [hunk 1]
```diff
@@ -21,28 +23,66 @@ import {
 import { t } from '../../../../../i18n';
 import RichText from '../../../../../components/RichText';
 
+import Joi from 'joi';
+
+const otherSourcesSchema = Joi.object({
```
Comment (thetif) [inline web/src/pages/apd/activities/cost-allocation/CostAllocation/OtherFunding.js:28]: can you move this to the schemas folder in common

File: web/src/pages/apd/activities/cost-allocation/CostAllocation/OtherFunding.js [hunk 2]
```diff
@@ -72,12 +134,30 @@ const OtherFunding = ({
...
+                  onChange={e => {
+                    setOtherFunding(activityIndex, ffy, value);;
+                    onChange(e);
+
+                    if (adminCheck) {
+                      trigger();
+                    }
+                  }}
+                  errorPlacement="bottom"
+                  errorMessage={errors?.costAllocation && errors?.costAllocation[ffy]?.other?.message}
```
Comment (thetif) [inline web/src/pages/apd/activities/cost-allocation/CostAllocation/OtherFunding.js:158]: you can switch this to `errors?.costAllocation?.[ffy]?.other?.message` and remove the `errors?.costAllocation &&`

File: web/src/pages/apd/activities/cost-allocation/CostAllocation/OtherFunding.js (hunk 1 above)
Comment (Sun-Mountain) [reply web/src/pages/apd/activities/cost-allocation/CostAllocation/OtherFunding.js:28]: Right!

**Comment by mirano-darren** [comment]
Nice work! (Reviewing for #3354, Dio is going to handle #3547)
| Given  | When | Then | Result |
| ------------- | ------------- | ------------- | ------------- |
| Description of cost allocation methodology field  | Changed to be blank/filled out  | If blank: error message appears. If filled out: error message doesn't appear  | PASS ✅  |
| Other funding amount field  | testing out of bounds numbers  | Anything less than 0 is invalid, letters and other symbols are also invalid  | PASS ✅   |
| Other funding amount field  | there is a validation error  | The validation error message is incorrect  |  FAIL ❌  |

image
Error message should be: Provide an other funding amount greater than or equal to $0.

**Review by mirano-darren — APPROVED** [review]
fixed! Going to wait on a review for #3547 and then i'll tag Jerome and Jenn

File: web/src/pages/apd/activities/cost-allocation/CostAllocation/OtherFunding.js (hunk 2 above)
Comment (thetif) [reply web/src/pages/apd/activities/cost-allocation/CostAllocation/OtherFunding.js:138]: it should be 
```
errorMessage={errors?.costAllocation?.[ffy]?.other?.message}
```

**Review by akuas — APPROVED** [review]
The description of cost allocation validation shows up when the field is left blank, as expected.

**Review by jeromeleecms — APPROVED** [review]
Looks to work as advertised. Fancy work on the valid other funding amount then forcing other funding description to validate. 

@beparticular do you need to review as well?

**Review by tbolt — APPROVED** [review]

**Pull request merged**